File: Year02/2nd semester/MPP/Lab/mpp-proiect-csharp-AlexandruPaul21/CsharpProj/repository/ClientRepository.cs
using System.Configuration;
using CsharpProj.model;
using log4net;
using Npgsql;

namespace CsharpProj.repository;

public class ClientRepository : IClientRepository
{
    //private readonly Logger logger = new Logger();
    
    private static readonly ILog logger = LogManager.GetLogger("ClientRepository");
    
    public Client findOne(long Id)
    {
        logger.Info("Client - findOne - Enter");
        Client client = new Client("", "");

        var sql = "SELECT * FROM clients WHERE id = @id";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@id", Id);
                var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    long id = reader.GetInt32(0);
                    var name = reader.GetString(1);
                    var address = reader.GetString(2);

                    client = new Client(name, address);
                    client.Id = id;
                }
            }
        }
        logger.Info("Client - findOne - Exit");
        return client;
    }

    public IEnumerable<Client> findAll()
    {
        logger.Info("Client - findAll - Enter");
        List<Client> clients = new List<Client>();
        var sql = "SELECT * FROM clients";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    long id = reader.GetInt32(0);
                    var name = reader.GetString(1);
                    var address = reader.GetString(2);

                    Client client = new Client(name, address);
                    client.Id = id;
                    
                    clients.Add(client);
                }
            }
        }
        logger.Info("Client - findOne - Exit");
        return clients;
    }

    public Client save(Client entity)
    {
        logger.Info("Client - save - Enter");
        var sql = "INSERT INTO clients VALUES (@id, @client_name, @address)";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@id", entity.Id);
                cmd.Parameters.AddWithValue("@client_name", entity.Name);
                cmd.Parameters.AddWithValue("@address", entity.Address);

                int rowsAff = cmd.ExecuteNonQuery();

                if (rowsAff != 0)
                {
                    entity = null!;
                }
            }
        }
        logger.Info("Client - save - Exit");
        return entity;
    }

    public long getLowestAvailableId()
    {
        logger.Info("Client - getLowestAvailableId - Enter");
        long id = 0;
        var sql = "SELECT id FROM clients ORDER BY id";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    long id1 = reader.GetInt32(0);
                    if (id1 != id)
                    {
                        break;
                    }
                    id++;
                }
            }
        }
        logger.Info("Client - getLowestAvailableId - Exit");
        return id;
    }
}

File: Year02/2nd semester/MPP/Lab/mpp-proiect-java-AlexandruPaul21/JavaProjectNetworking/ProjectPersistance/src/main/java/org/example/repository/jdbc/ClientRepository.java
package org.example.repository.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.Client;
import org.example.repository.IClientRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ClientRepository implements IClientRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public ClientRepository(Properties properties) {
        logger.info("Initialising ClientRepository with properties {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    public Long getLowestAvbId() {
        String sql = "SELECT MAX(id) FROM clients";
        Long id = null;

        try (
                Connection connection = jdbcUtils.getConnection();
                PreparedStatement preparedStatement = connection.prepareStatement(sql)
                ) {

            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                id = resultSet.getLong(1);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return id;
    }

    @Override
    public Client findOne(Long aLong) {
        logger.traceEntry();
        String sql = "SELECT * FROM clients WHERE id = ?";

        Client client = null;

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setLong(1, aLong);

            ResultSet resultSet = preparedStatement.executeQuery();

            if (resultSet.next()) {
                String clientName = resultSet.getString("client_name");
                String address = resultSet.getString("address");

                client = new Client(clientName, address);
                client.setId(aLong);
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }

        logger.traceExit();
        return client;
    }

    @Override
    public Iterable<Client> findAll() {
        logger.traceEntry();
        String sql = "SELECT * FROM clients";

        List<Client> clients = new ArrayList<>();

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);

            ResultSet resultSet = preparedStatement.executeQuery();

            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String clientName = resultSet.getString("client_name");
                String address = resultSet.getString("address");

                Client client = new Client(clientName, address);
                client.setId(id);

                clients.add(client);
            }

        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }

        logger.traceExit();
        return clients;
    }

    @Override
    public Client save(Client entity) {
        String sql = "INSERT INTO clients VALUES (?, ?, ?)";

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);

            preparedStatement.setLong(1, entity.getId());
            preparedStatement.setString(2, entity.getName());
            preparedStatement.setString(3, entity.getAddress());

            preparedStatement.executeUpdate();
            entity = null;
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return entity;
    }
}
